src/advuilist.h

```diff
@@ -0,0 +1,889 @@
+    // reset exit variable in case suspend() was called earlier
+    _exit = false;
+
+    if( !_ui ) {
+        _initui();
+    }
+
+    while( !_exit ) {
+
+        ui_manager::redraw();
+        std::string const action = _ctxt.handle_input();
+
+        if( action == ACTION_UP ) {
+            _decidx( 1 );
+        } else if( action == ACTION_DOWN ) {
+            _incidx( 1 );
+        } else if( action == ACTION_PAGE_UP ) {
+            _decidx( _pagesize );
+        } else if( action == ACTION_PAGE_DOWN ) {
+            _incidx( _pagesize );
+        } else if( action == ACTION_SORT ) {
+            _querysort();
+        } else if( action == ACTION_FILTER ) {
+            _queryfilter();
+        } else if( action == ACTION_RESET_FILTER ) {
+            _setfilter( std::string() );
+        } else if( action == ACTION_SELECT ) {
+            return peek();
+        } else if( action == ACTION_SELECT_PARTIAL ) {
+            std::size_t const count = _peekcount();
+            std::size_t const input = _querypartial( count );
+            if( input > 0 ) {
+                return _peek( input );
+            }
+        } else if( action == ACTION_SELECT_WHOLE ) {
+            std::size_t const count = _peekcount();
+            return _peek( count );
+        } else if( action == ACTION_SELECT_ALL ) {
+            return _peekall();
+        } else if( action == ACTION_QUIT ) {
+            _exit = true;
+        }
+
+        if( _fctxt ) {
+            _fctxt( this, action );
+        }
+    }
+
+    return select_t();
+}
+
+template <class Container, typename T>
+void advuilist<Container, T>::sort( std::string const &name )
+{
+    auto const it = std::find_if( _sorters.begin(), _sorters.end(), [&]( sorter_t const & v ) {
+        return std::get<std::string>( v ) == name;
+    } );
+    if( it != _sorters.end() ) {
+        _sort(
+            static_cast<typename sortcont_t::size_type>( std::distance( _sorters.begin(), it ) ) );
+    }
+}
+
+template <class Container, typename T>
+void advuilist<Container, T>::rebuild( Container *list )
+{
+    _list.clear();
+    Container *rlist = list == nullptr ? _olist : list;
+    std::size_t idx = 0;
+    for( auto it = rlist->begin(); it != rlist->end(); ++it ) {
+        if( !_ffilter or _filter.empty() or _ffilter( *it, _filter ) ) {
+            if( _frebuild ) {
+                _frebuild( _list.empty(), *it );
+            }
+            _list.emplace_back( idx++, it );
+        }
+    }
+    _group( _cgroup );
+    _sort( _csort );
+    _setidx( _cidx );
+    _cpage = _idxtopage( _cidx );
+}
+
+template <class Container, typename T>
+typename advuilist<Container, T>::select_t advuilist<Container, T>::peek()
+{
+    return _peek( 1 );
+}
+
+template <class Container, typename T>
+void advuilist<Container, T>::suspend()
+{
+    _exit = true;
+}
+
+template <class Container, typename T>
+void advuilist<Container, T>::totop()
+{
+    // FIXME: is there a better way? Ask Qrox
+    _initui();
```

Comment:
I didn't add the functionality to change z-order of `ui_adaptor` because the need did not come up when I was migrating the old code to `ui_adaptor`. Is this needed here because some of the UIs overlap?

Reply:
That was originally needed for switching panes in transaction_ui where my active pane wasn't redrawing unless it was on top. Don't think that I've overlapped them but I'll look into it. In any case `totop()` has effectively become an `init_ui()` since 66a3eb0dd055cd3b8949996f3730986d5f7df59e where I delayed the creation of `ui_adaptor`

Reply:
> my active pane wasn't redrawing unless it was on top.

You can call `ui_adaptor::invalidate_ui` so the UI is always redrawn the next time `ui_manager::redraw` is called. (`ui_manager::redraw` is actually just a shorthand for invalidating the topmost UI and then calling `ui_manager::redraw_invalidated`).

Reply:
Wouldn't that cause both panes to get redrawn if both I invalidate and call `ui_manager::redraw` in `advuilsit::select()`? Take a quick look at `transaction_ui::_process()` to see where I need to call `totop()` (in other cases it's just for initializing the ui).

Reply:
You can invalidate and call `ui_manager::redraw_invalidated`, so only the invalidated UI will be redrawn (as long as the UIs do not overlap).

Reply:
Thanks! I've addressed this in ec20c38. Let me know if there are still issues.